kernel/thread/worker.c

```diff
@@ -0,0 +1,89 @@
+/* KallistiOS ##version##
+
+   worker.c
+   Copyright (C) 2024 Paul Cercueil
+*/
+
+#include <kos/genwait.h>
+#include <kos/thread.h>
+#include <kos/worker_thread.h>
+#include <stdbool.h>
+#include <stdlib.h>
+
+struct kthread_worker {
+    kthread_t *thd;
+    void (*routine)(void *);
+    void *data;
+    bool pending;
+    bool quit;
+};
+
+static void * thd_worker_thread(void *d) {
+    kthread_worker_t *worker = d;
+    uint32_t flags;
+
+    for (;;) {
+        flags = irq_disable();
+
+        if (!worker->pending)
+            genwait_wait(worker, worker->thd->label, 0, NULL);
+
+        irq_restore(flags);
+
+        if (worker->quit)
+            break;
+
+        worker->pending = false;
+        worker->routine(worker->data);
+    }
+
+    return NULL;
+}
+
+kthread_worker_t * thd_worker_create(void (*routine)(void *), void *param) {
+    kthread_worker_t *worker;
+    uint32_t flags;
+
+    worker = malloc(sizeof(*worker));
+    if (!worker)
+        return NULL;
+
+    worker->data = param;
+    worker->routine = routine;
+    worker->pending = false;
+    worker->quit = false;
+
+    flags = irq_disable();
+
+    worker->thd = thd_create(0, thd_worker_thread, worker);
+    if (!worker->thd) {
+        irq_restore(flags);
+        free(worker);
+        return NULL;
+    }
+
+    irq_restore(flags);
+
+    return worker;
+}
+
+void thd_worker_wakeup(kthread_worker_t *worker) {
+    uint32_t flags = irq_disable();
+
+    worker->pending = true;
+    genwait_wake_one(worker);
```

Comment:
What happens if the worker thread just hasn't run yet in order to sleep on itself? The way I read it the genwait_wake_one will fail, but now the pending will be set to true so whenever the thread does run for the first time it won't sleep but just run through? Makes sense, but just wanted to verify my reading.

That led me to the issue in that calling `thd_worker_wakeup` X times does not guarantee that the worker routine ran X times. I'm not sure if that is intentional, but it would certainly seem to me to limit the usefulness. 

Changing `pending` to an int rather than a bool could fix that and generally improve the usability of the system, but otherwise I'd think that this wakeup would need to at least be able to signal to the caller of it if the worker was already awake (to know that calling the function would have no effect).

Reply:
> What happens if the worker thread just hasn't run yet in order to sleep on itself? The way I read it the genwait_wake_one will fail, but now the pending will be set to true so whenever the thread does run for the first time it won't sleep but just run through? Makes sense, but just wanted to verify my reading.

Correct.

The "pending" is used when `thd_worker_wakeup` is called while the worker thread is running; it tells it that more work was added while the thread was running, so the worker should restart when it's done with the current load.
 
> That led me to the issue in that calling `thd_worker_wakeup` X times does not guarantee that the worker routine ran X times. I'm not sure if that is intentional, but it would certainly seem to me to limit the usefulness.

It is not an issue, it is by design. The wakeup function basically says "there is work to do", it doesn't actually give your worker function any work load - this is completely up to you how you implement this.

See it for instance in the context of a IRQ handler: when the worker function start, it would read the "pending IRQ" register, cache it into a variable, ACK the interrupts, and then for each bit set, handle the corresponding interrupt. It wouldn't handle just one single interrupt then return, but handle all the ones that were pending. And if an IRQ fires while the worker thread is running, `thd_worker_wakeup` would be called, which would cause the worker thread to restart the IRQs processing after it's done with the current batch.

What you are asking for is trivial to implement on top of the existing API, by the way; you'd just need to implement a FIFO for workloads. Wakeup the worker when something was added, and the worker thread processes the jobs from the FIFO until it's empty.

I can totally add an API for that, but my point is, the current one is fine as-is.

> 
> Changing `pending` to an int rather than a bool could fix that and generally improve the usability of the system, but otherwise I'd think that this wakeup would need to at least be able to signal to the caller of it if the worker was already awake (to know that calling the function would have no effect).

I don't see why the caller needs to know anything about that.